djinni/cwrapper/wrapper_marshal.hpp

```diff
@@ -0,0 +1,219 @@
+// preliminary
+// Part of support library for djinni4python, handwritten
+
+#pragma once
+#include <string>
+#include <vector>
+#include <cstdint>
+#include <chrono>
+#include <thread>
+#include <assert.h>
+#include "../../test-suite/handwritten-src/cpp/optional.hpp"
```

Comment:
Oops, whats that? I don't think the support-lib should depend on a header from the test-suite 😳

Reply:
I've been trying to fix this the last couple of days and the only way I made it work was by bumping the C++ standard to C++17, which I believe @a4z would be ok with. I also had to change 3 places in the generator where `std::experimental::...` was hardcoded. That was I was able to use `std::optional`. Should we go that way?

Reply:
For other languages, Djinni has historically taken a command-line argument specifying how to include the `optional` header, so as to support C++ versions from before it was standard.  That was one of the missing features in Python when it released in beta, though I didn't remember that this was the hack to get around it.  This line could probably be updated to depend on a generated header which depends on the command-line arg, but it may also be that `optional` has been in the standard library long enough to make it a minimal requirement.

Reply:
IMHO, for maintenance purpose and keep the project alive, I am all for simplify what ever is possible, and I think that means going to c++17.

All main target platforms have support for that since years, and we simply do not have resources to make things more complicated than required and support theoretical use-cases no one of us has (anymore). Not even mentioning that all that options are barely tested, if at all.

If that means we lose some existing user in new versions who can not update compiler flags for reasons, they are fine with existing versions since they work.

Reply:
How should we go about this?

We could use the command-line parameter to specify which optional type to use and, for some combination of C++ standard, target languages, and/or platform, only one or another optional type would apply. However, the Python bridge uses another optional-related type, `std::[experimental::]nullptr`, which is not parameterized from the Djinni command-line.

If we opt for not using the command-line parameter, we still have two options for this PR:

1. Use `std::optional` and `std::nullopt` just where it's required for the Python bridge to work, or
2. Retire the command-line parameters for optional at once and the Python PR waits until it's done.

_Note: no matter what we chose, I'm afraid we'll need another generator release because of the hardcoded `std::experimental...` code._

Reply:
No problem of making one more generator release, if that is required, than it is.
We are in the process of figuring out on how to handle that dependency, it should  not be that hard to solve, the main issue is time to investigate on that, not a technical one.